statistics/cmsketch.go

```diff
@@ -44,29 +57,208 @@ func NewCMSketch(d, w int32) *CMSketch {
+
+	for _, sl := range counter {
+		for i := range sl {
+			if enableTopN && sl[i].count >= topnCountThreshold {
+				topn = append(topn, cmscount{data: sl[i].data, count: sl[i].count * ratio})
+			} else {
+				c.InsertBytesN(sl[i].data, sl[i].count*ratio)
+			}
+		}
+	}
+
+	if enableTopN {
+		c.buildTopNMap(topn)
+	}
+
+	// Interesting, seems we have collected all distinct values.
+	// These three tests tests if all divisions are legal.
+	// They also tests if we sampled all possible data.
+	countWithoutTopN := total - (sampleSize-uint64(onlyOnceItems))*ratio
+	if total < topncount*ratio {
+		c.defaultValue = 1
+	} else if estimateNDV <= uint64(realTopN) {
+		c.defaultValue = 1
+	} else {
+		c.defaultValue = countWithoutTopN / (estimateNDV - uint64(ndv) + onlyOnceItems)
+	}
+}
+
+func (c *CMSketch) buildTopNMap(topn []cmscount) {
+	c.topnindex = make(map[uint64][]cmscount)
+	for i := range topn {
+		if topn[i].data == nil {
+			continue
+		}
+		h1, h2 := murmur3.Sum128(topn[i].data)
+		sl, ok := c.topnindex[h1]
+		if !ok {
+			sl = make([]cmscount, 0)
+		}
+		sl = append(sl, cmscount{h1, h2, topn[i].data, topn[i].count})
+		c.topnindex[h1] = sl
+	}
+}
+
+// queryAddTopN TopN adds count to CMSketch.topnindex if exists, and returns the count of such elements after insert
+// if such elements does not in topn elements, nothing will happen and false will be returned.
+func (c *CMSketch) queryAddTopN(h1, h2, count uint64, d []byte) (uint64, bool) {
+	if c.topnindex == nil {
+		return 0, false
+	}
+	cnt, ok := c.topnindex[h1]
+	if !ok {
+		return 0, false
+	}
+	for k := range cnt {
+		if cnt[k].h2 == h2 && bytes.Equal(d, cnt[k].data) {
+			cnt[k].count += count
+			return cnt[k].count, true
+		}
+	}
+	return 0, false
+}
+
+func (c *CMSketch) queryTopN(h1, h2 uint64, d []byte) (uint64, bool) {
+	return c.queryAddTopN(h1, h2, 0, d)
+}
+
 // InsertBytes inserts the bytes value into the CM Sketch.
 func (c *CMSketch) InsertBytes(bytes []byte) {
-	c.count++
+	c.InsertBytesN(bytes, 1)
+}
+
+// InsertBytesN adds the bytes value into the CM Sketch by n.
+func (c *CMSketch) InsertBytesN(bytes []byte, n uint64) {
 	h1, h2 := murmur3.Sum128(bytes)
+	if _, ok := c.queryAddTopN(h1, h2, n, bytes); ok {
+		return
+	}
+	c.count += n
 	for i := range c.table {
 		j := (h1 + h2*uint64(i)) % uint64(c.width)
-		c.table[i][j]++
+		c.table[i][j] += uint32(n)
 	}
 }
 
 // setValue sets the count for value that hashed into (h1, h2).
 func (c *CMSketch) setValue(h1, h2 uint64, count uint32) {
 	oriCount := c.queryHashValue(h1, h2)
-	c.count += uint64(count) - uint64(oriCount)
+
+	if oriCount < 2*(c.count/uint64(c.width)) && c.defaultValue > 0 {
+		// This case, we should also update c.defaultValue
+		// Set default value directly will result in more error, instead, update it by 5%.
+		// This should make estimate better, if defaultValue becomes 0 frequently, commit this line.
+		c.defaultValue = uint64(float64(c.defaultValue)*0.95 + float64(c.defaultValue)*0.05 + 0.5)
```

Comment:
I'm confused by this line. When will it update since it almost equals to `c.defaultValue+0.5`?